builtin/common/strict.lua

```diff
@@ -14,33 +14,34 @@ local declared = {}
 local warned = {}
 
 function meta:__newindex(name, value)
+	if declared[name] then
+		return
+	end
 	local info = getinfo(2, "Sl")
 	local desc = ("%s:%d"):format(info.short_src, info.currentline)
-	if not declared[name] then
-		local warn_key = ("%s\0%d\0%s"):format(info.source,
-				info.currentline, name)
-		if not warned[warn_key] and info.what ~= "main" and
-				info.what ~= "C" then
-			core.log("warning", ("Assignment to undeclared "..
-					"global %q inside a function at %s.")
+	local warn_key = ("%s\0%d\0%s"):format(info.source, info.currentline, name)
+	if not warned[warn_key] and info.what ~= "main" and info.what ~= "C" then
+		core.log("warning", ("Assignment to undeclared global %q inside a function at %s.")
 				:format(name, desc))
-			warned[warn_key] = true
-		end
-		declared[name] = true
+		warned[warn_key] = true
 	end
 	rawset(self, name, value)
+	declared[name] = true
 end
 
 
 function meta:__index(name)
+	if declared[name] then
+		return nil
+	end
 	local info = getinfo(2, "Sl")
 	local warn_key = ("%s\0%d\0%s"):format(info.source, info.currentline, name)
-	if not declared[name] and not warned[warn_key] and info.what ~= "C" then
+	if not warned[warn_key] and info.what ~= "C" then
 		core.log("warning", ("Undeclared global variable %q accessed at %s:%s")
 				:format(name, info.short_src, info.currentline))
 		warned[warn_key] = true
 	end
-	return rawget(self, name)
+	return nil
```

Comment:
`__index` is only called if the value is `nil`, so there's no need to `rawget` here